src/v2/providers/storage.ts

```diff
@@ -0,0 +1,419 @@
+      event: CloudEvent<StorageObjectData>
+    ) => any | Promise<any>;
+    _onOperation(handler, deletedEvent);
+  }
+  return _onOperation(
+    handler,
+    deletedEvent,
+    optsOrHandler as string | StorageOptions
+  );
+}
+
+/** Handle a storage object metadata updated */
+export function onObjectMetadataUpdated(
+  handler: (event: CloudEvent<StorageObjectData>) => any | Promise<any>
+): CloudFunction<StorageObjectData>;
+
+/** Handle a storage object metadata updated */
+export function onObjectMetadataUpdated(
+  bucket: string,
+  handler: (event: CloudEvent<StorageObjectData>) => any | Promise<any>
+): CloudFunction<StorageObjectData>;
+
+/** Handle a storage object metadata updated */
+export function onObjectMetadataUpdated(
+  opts: StorageOptions,
+  handler: (event: CloudEvent<StorageObjectData>) => any | Promise<any>
+): CloudFunction<StorageObjectData>;
+
+export function onObjectMetadataUpdated(
+  optsOrHandler:
+    | string
+    | StorageOptions
+    | ((event: CloudEvent<StorageObjectData>) => any | Promise<any>),
+  handler?: (event: CloudEvent<StorageObjectData>) => any | Promise<any>
+): CloudFunction<StorageObjectData> {
+  if (arguments.length === 1) {
+    handler = optsOrHandler as (
+      event: CloudEvent<StorageObjectData>
+    ) => any | Promise<any>;
+    _onOperation(handler, metadataUpdatedEvent);
+  }
+  return _onOperation(
+    handler,
+    metadataUpdatedEvent,
+    optsOrHandler as string | StorageOptions
+  );
+}
+
+/** @hidden */
+export function _onOperation(
+  handler: (event: CloudEvent<StorageObjectData>) => any | Promise<any>,
+  eventType: string,
+  bucketOrOpts?: string | StorageOptions
+): CloudFunction<StorageObjectData> {
+  const [opts, bucket] = _getOptsAndBucket(bucketOrOpts);
+
+  const func = (raw: CloudEvent<unknown>) => {
+    return handler(raw as CloudEvent<StorageObjectData>);
+  };
+
+  func.run = handler;
+
+  // TypeScript doesn't recongize defineProperty as adding a property and complains
+  // that __trigger doesn't exist. We can either cast to any and lose all type safety
+  // or we can just assign a meaningless value before calling defineProperty.
+  func.__trigger = 'silence the transpiler';
+
+  Object.defineProperty(func, '__trigger', {
+    get: () => {
+      const baseOpts = options.optionsToTriggerAnnotations(
+        options.getGlobalOptions()
+      );
+      const specificOpts = options.optionsToTriggerAnnotations(opts);
+
+      return {
+        // TODO(colerogers): Remove "apiVersion" once the CLI has migrated to
+        // "platform"
+        apiVersion: 2,
+        platform: 'gcfv2',
+        ...baseOpts,
+        ...specificOpts,
+        labels: {
+          ...baseOpts?.labels,
+          ...specificOpts?.labels,
+        },
+        eventTrigger: {
+          eventType: provider + '.' + eventType,
+          resource: bucket, // TODO(colerogers): replace with bucket: 'my-bucket' when container contract is finished
```

Comment:
Can you expand on what you mean by this TODO? I'm a confused by the suggestion that we should set this to a constant `my-bucket`. Also probably better to leave out mentioning container contract here.

Reply:
Good catch, I removed the CC part. Eventually, we'll need to remove the resource field and just map to a bucket field.